frappe/tests/test_document.py

```diff
@@ -98,6 +98,13 @@ def test_update(self):
 
 		self.assertEqual(frappe.db.get_value(d.doctype, d.name, "subject"), "subject changed")
 
+	def test_discard(self):
+		d = self.test_insert()
+		self.assertEqual(d.docstatus, 0)
+
+		d.discard()
+		self.assertEqual(d.docstatus, 2)
+
```

Comment:
ummm, add more tests. Think of all the use cases / transitions you're promising with this feature.

Reply:
- cancel/submit -> discard shouldn't work
- discard to draft shouldn't work
- discard shouldn't trigger any on_cancel events
- discard shouldn't trigger save of entire doc, if any code does it from events it should cause a validation failure (cause save might endup triggering cancel events)
- `_action=discard` should be asserted and usage be checked in code

Reply:
We can drop the _action change because I'm doing transition checks explicitly in discard

```python		
if not self.docstatus == DocStatus.draft():
    raise frappe.ValidationError(_("Only draft documents can be discarded"), self.docstatus)
```

Doing this because of some conflicting rules that should be invalid in other scenarios like,
0 > 2 - should not be allowed for draft.cancel() but allowed for draft.discard()
1 > 2 - should not be allowed for submitted.discard() but allowed for submitted.cancel()